### PR TITLE
ARROW-8795: [C++] Limited iOS support

### PR DESCRIPTION
**Pull request by ilijapuaca**
Per suggestion from [JIRA](https://issues.apache.org/jira/browse/ARROW-8795), opening up a pull request to potentially get the ball rolling on having iOS as one of the supported platforms.

Per writeup in [this doc](https://github.com/UnfoldedInc/deck.gl-native-dependencies/blob/master/docs/iOS-BUILD.md#arrow-v0170), there are a few issues with cmake's Xcode generator that require a few workarounds even when using their most recent versions.

The writeup also describes issues that we've encountered building the library, as well as issues that were not resolved as they use optional features we are currently not making use of.

### CODE REVIEW

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -189,6 +189,7 @@ set(ARROW_SRCS
     util/utf8.cc
     util/value_parsing.cc
     vendored/base64.cpp
+    vendored/datetime/ios.mm
```
Comment (pitrou) [inline cpp/src/arrow/CMakeLists.txt:192]: Shouldn't this file only be compiled on iOS targets? At least it seems it doesn't compile properly on our CI builds.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -340,6 +340,11 @@ if(ARROW_ORC)
   set(ARROW_WITH_ZSTD ON)
 endif()
 
+# datetime code used by iOS requries zlib support
```
Comment (pitrou) [inline cpp/CMakeLists.txt:343]: Typo: "requires"

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for doing this. On the principle this looks good, but see below.

File: cpp/src/arrow/CMakeLists.txt (hunk 1 above)
Comment (ilijapuaca) [reply cpp/src/arrow/CMakeLists.txt:192]: I wasn't sure about usage of these files in general -- `ios.h` is included from `tz.cpp` under `#ifdef __APPLE__`, which implies it's being used on macOS as well?

I could make this file iOS-specific, but perhaps some additional updates should be made in order to consolidate `datetime` module and its usage. If that is the case, I'd need some pointers as I'm not familiar with this portion of the codebase and its usage

Comment (pitrou) [reply cpp/src/arrow/CMakeLists.txt:192]: Hmm, then at least make this Apple-specific. `ios.mm` needs an Objective-C++ compiler, which won't be found on all platforms.

**Comment by ilijapuaca** [comment]
@pitrou this now seems to build on all the platforms. Tried running `run-cmake-format.py` to fix the lint issue, but for some reason it produces a much different different format across all the cmake files on my machine, which I assume are not formatted correctly.

Any idea why that might be the case? If this seems mergeable now, perhaps someone else can fix the linter issue if this isn't a known issue that I can easily solve on my machine

**Comment by pitrou** [comment]
@ilijapuaca Ah, sorry for that. You need a specific `cmake-format` version:
```
pip install cmake-format==0.5.2
```
should do the trick.

**Comment by ilijapuaca** [comment]
> @ilijapuaca Ah, sorry for that. You need a specific `cmake-format` version:
> 
> ```
> pip install cmake-format==0.5.2
> ```
> 
> should do the trick.

@pitrou Yep, that worked much better. Pushed the formatting update

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**